### PR TITLE
Feature no underscore: removed underscore js dependency

### PR DESCRIPTION
**Pull request by gangadharjannu**
created two utility functions
1. forEach for converting iterating over node list.
2. getType for fixing typeof and getting correct type of variable.

Replaced below methods:
`_.isObject, _.isString, _.isNumber, _.isBoolean, _.isNull, _.isArray` with custom getType method
`_.each` with [Array.prototype.forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
`_.filter` with [Array.prototype.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)
`_.find` with [Array.prototype.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)
`_.findIndex` with [Array.prototype.findIndex](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex)
`_.indexOf` with [Array.prototype.indexOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf)

Fixed lintintg issues
code indentation according to visual studio code.
created two utility functions
1. forEach for converting iterating over node list.
2. getType for fixing typeof and getting correct type of variable.

Replaced below methods:
`_.isObject, _.isString, _.isNumber, _.isBoolean, _.isNull, _.isArray` with custom getType method
`_.each` with [Array.prototype.forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
`_.filter` with [Array.prototype.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)
`_.find` with [Array.prototype.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)
`_.findIndex` with [Array.prototype.findIndex](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex)
`_.indexOf` with [Array.prototype.indexOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf)

Fixed lintintg issues
code indentation according to visual studio code.
Added polyfills for Array.prototype.find and Array.prototype.findIndex

### CODE REVIEW

**Comment by gangadharjannu** [comment]
HI @zgrossbart 
I have removed the dependency of underscore.js which weighs around 20Kb. Yes, I would say this as refactoring.

**Comment by zgrossbart** [comment]
Saving 20 kb is great, but it looks like you made a lot more changes than that.  Can you please let me know more about the changes you made.  Also, what browsers have you tested on?  What about older version?

**Comment by gangadharjannu** [comment]
Before answering your question. Can you tell me the current browser support?

**Comment by zgrossbart** [comment]
I'm trying to support Chrome latest, Safari latest, Firefox latest, IE, and Edge.  I'm also trying to make sure it works well on all mobile devices.

File: jdd.js [hunk 1]
```diff
@@ -1088,3 +1085,130 @@ jQuery(document).ready(function() {
...
+    });
+}
+
+// Array.prototype.findIndex
+// https://tc39.github.io/ecma262/#sec-array.prototype.findIndex
+if (!Array.prototype.findIndex) {
+    Object.defineProperty(Array.prototype, 'findIndex', {
+        value: function (predicate) {
+            // 1. Let O be ? ToObject(this value).
+            if (this == null) {
```
Comment (zgrossbart) [inline jdd.js]: Should this be triple equals and no double?

File: jdd.js [hunk 2]
```diff
@@ -1088,3 +1085,130 @@ jQuery(document).ready(function() {
...
+                throw new TypeError('"this" is null or not defined');
+            }
+
+            var o = Object(this);
+
+            // 2. Let len be ? ToLength(? Get(O, "length")).
+            var len = o.length >>> 0;
```
Comment (zgrossbart) [inline jdd.js]: Why are you using >>> to compare length?

File: jdd.js [hunk 3]
```diff
@@ -1088,3 +1085,130 @@ jQuery(document).ready(function() {
...
```
Comment (zgrossbart) [inline jdd.js]: Why are you using >>> to compare length?

File: jdd.js [hunk 4]
```diff
@@ -1088,3 +1085,130 @@ jQuery(document).ready(function() {
...
+
+// polyfills
+
+// Array.prototype.find
+// https://tc39.github.io/ecma262/#sec-array.prototype.find
+if (!Array.prototype.find) {
+    Object.defineProperty(Array.prototype, 'find', {
+        value: function (predicate) {
+            // 1. Let O be ? ToObject(this value).
+            if (this == null) {
```
Comment (zgrossbart) [inline jdd.js]: Shouldn't this be === instead of ==?

File: jdd.js [hunk 5]
```diff
@@ -1088,3 +1085,130 @@ jQuery(document).ready(function() {
...
+}
+/**
+ * Iterate over array of objects and call given callback for each item in the array
+ * Optionally may take this as scope
+ * 
+ * @param array 
+ * @param callback 
+ * @param optional scope 
+ */
+function forEach(array, callback, scope) {
```
Comment (zgrossbart) [inline jdd.js]: Why do we need a custom `forEach` function and why are we defining it so far down in the file when it's used higher up?

File: jdd.js [hunk 6]
```diff
@@ -1088,3 +1085,130 @@ jQuery(document).ready(function() {
...
+
+// utilites
+// 
+/**
+ * Fixing typeof
+ * takes value and returns type of value
+ * @param  value 
+ * return typeof value
+ */
+function getType(value) {
```
Comment (zgrossbart) [inline jdd.js]: Why are we defining functions like this in the global scope and not on the `jdd` object?

File: jdd.js (hunk 5 above)
Comment (gangadharjannu) [reply jdd.js]: Array.prototype.forEach only support iterating over arrays however we need to iterate over nodeList which is an array like structure. Array.prototype.forEach doesn't support iterating over nodeList. So I've used a custom method. I could've used Array.prototype.forEach.call however I already have gone through this https://toddmotto.com/ditch-the-array-foreach-call-nodelist-hack/. So I have crated my own custom forEach function which enables iterating over array like objects.

File: jdd.js (hunk 3 above)
Comment (gangadharjannu) [reply jdd.js]: It is the standard implementation across mozilla for converting length to unsigned 32 bit integer. https://stackoverflow.com/a/3082073/3543808

File: jdd.js (hunk 2 above)
Comment (gangadharjannu) [reply jdd.js]: It is the standard implementation across mozilla for converting length to unsigned 32 bit integer. https://stackoverflow.com/a/3082073/3543808

File: jdd.js (hunk 4 above)
Comment (gangadharjannu) [reply jdd.js]: for checking null/undefined values the above comparison should be suffice. https://softwareengineering.stackexchange.com/a/268125/258611

File: jdd.js (hunk 1 above)
Comment (gangadharjannu) [reply jdd.js]: for checking null/undefined values the above comparison should be suffice. https://softwareengineering.stackexchange.com/a/268125/

**Comment by gangadharjannu** [comment]
Mostly the changes that you suggested are from rock solid polyfills provided by Mozilla. The other change regarding jdd namespace is feasible and in fact I wanted to do so. However as I explained in my comment I defined outside just for separating concerns.

**Pull request merged**